src/ppl/src/SimulatedAnnealing.cpp

```diff
@@ -0,0 +1,206 @@
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "SimulatedAnnealing.h"
+
+#include "utl/Logger.h"
+#include "utl/algorithms.h"
+
+namespace ppl {
+
+SimulatedAnnealing::SimulatedAnnealing(Netlist* netlist,
+                                       std::vector<Slot>& slots,
+                                       Logger* logger,
+                                       odb::dbDatabase* db)
+    : netlist_(netlist), slots_(slots), logger_(logger), db_(db)
+{
+  num_slots_ = slots.size();
+  num_pins_ = netlist->numIOPins();
+  perturb_per_iter_ = static_cast<int>(num_pins_ * 0.8);
+}
+
+void SimulatedAnnealing::init()
+{
+  pin_assignment_.resize(num_pins_);
+  slot_indices_.resize(num_slots_);
+  std::iota(slot_indices_.begin(), slot_indices_.end(), 0);
+
+  std::mt19937 rand_gen(seed_);
+  generator_ = rand_gen;
+  std::uniform_real_distribution<float> distribution(0.0, 1.0);
+  distribution_ = distribution;
+}
+
+void SimulatedAnnealing::run()
+{
+  init();
+  randomAssignment();
+  int pre_cost = getAssignmentCost();
+  float temperature = init_temperature_;
+
+  for (int iter = 0; iter < max_iterations_; iter++) {
+    for (int perturb = 0; perturb < perturb_per_iter_; perturb++) {
+      int prev_slot = -1;
+      int new_slot = -1;
+      int pin1 = -1;
+      int pin2 = -1;
+      perturbAssignment(prev_slot, new_slot, pin1, pin2);
+      int cost = getAssignmentCost();
```

Comment:
Recomputing the assignment cost visits every net.  When you perturb you only touch one or two nets.   It would be faster to compute a cost delta for the move.